File: node-js-deposit-redeem-example/price.js
require("dotenv").config();
const { ethers } = require("ethers");

const rwaAbi = require("./abis/RWAToken.json").abi;
const feeAbi = require("./abis/RWAFee.json").abi;

/**
 * Read-only information script.
 *
 * Purpose:
 * - show deposit preview information
 * - show redeem preview information
 * - show convertToAssets information
 *
 * Usage:
 *   node price.js [desiredAssetsBeforeFee] [shares]
 *
 * Example:
 *   node price.js 100 25
 */
function getEnv() {
  const env = {
    RPC_URL: process.env.RPC_URL,
    RWA_TOKEN_ADDRESS: process.env.RWA_TOKEN_ADDRESS,
    FEE_ADDRESS: process.env.FEE_ADDRESS || "",
    ENTRY_FEE_TIMING: Number(process.env.ENTRY_FEE_TIMING || 0),
    EXIT_FEE_TIMING: Number(process.env.EXIT_FEE_TIMING || 1),
  };

  if (!env.RPC_URL || !env.RWA_TOKEN_ADDRESS) {
    throw new Error("Missing required env values. Please check .env.example");
  }

  return env;
}

async function resolveFeeAddress(vault, env) {
  if (env.FEE_ADDRESS) return env.FEE_ADDRESS;

  const feeAddress = await vault.rwaFee();
  if (!feeAddress || feeAddress === ethers.ZeroAddress) {
    throw new Error("Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address.");
  }

  return feeAddress;
}

async function main() {
  const desiredAssetsInput = process.argv[2] || "100";
  const sharesInput = process.argv[3] || "25";

  const env = getEnv();

  const provider = new ethers.JsonRpcProvider(env.RPC_URL);
  const vault = new ethers.Contract(env.RWA_TOKEN_ADDRESS, rwaAbi, provider);

  const assetAddress = await vault.asset();
  const asset = new ethers.Contract(
    assetAddress,
    [
      "function decimals() view returns (uint8)",
      "function symbol() view returns (string)"
    ],
    provider
  );

  const shareDecimals = await vault.decimals();
  const shareSymbol = await vault.symbol();
  const assetDecimals = await asset.decimals();
  const assetSymbol = await asset.symbol();

  const feeAddress = await resolveFeeAddress(vault, env);
  const feeContract = new ethers.Contract(feeAddress, feeAbi, provider);

  const desiredAssets = ethers.parseUnits(desiredAssetsInput, assetDecimals);
  const shares = ethers.parseUnits(sharesInput, shareDecimals);

  // Deposit info: user enters desired assets before fee
  const entryFee = await feeContract.feeOnRaw(desiredAssets, env.ENTRY_FEE_TIMING);
  const totalSpend = desiredAssets + entryFee;
  const previewDepositShares = await vault.previewDeposit(totalSpend);

  // Redeem info: previewRedeem already represents final net
  const grossAssets = await vault.convertToAssets(shares);
  const estimatedExitFee = await feeContract.feeOnTotal(grossAssets, env.EXIT_FEE_TIMING);
  const previewRedeemNetAssets = await vault.previewRedeem(shares);

  console.log("=== PRICE / INFO FLOW ===");
  console.log("Vault:", env.RWA_TOKEN_ADDRESS);
  console.log("Fee contract:", feeAddress);
  console.log("");

  console.log("--- Deposit information ---");
  console.log("Desired assets before fee:", desiredAssetsInput, assetSymbol);
  console.log("Entry fee (feeOnRaw):", ethers.formatUnits(entryFee, assetDecimals), assetSymbol);
  console.log("Total spend sent to deposit():", ethers.formatUnits(totalSpend, assetDecimals), assetSymbol);
  console.log("previewDeposit(totalSpend):", ethers.formatUnits(previewDepositShares, shareDecimals), shareSymbol);
  console.log("");

  console.log("--- Redeem information ---");
  console.log("Input shares:", sharesInput, shareSymbol);
  console.log("convertToAssets(shares) [gross info]:", ethers.formatUnits(grossAssets, assetDecimals), assetSymbol);
  console.log("Estimated exit fee (feeOnTotal):", ethers.formatUnits(estimatedExitFee, assetDecimals), assetSymbol);
  console.log("previewRedeem(shares) [final net]:", ethers.formatUnits(previewRedeemNetAssets, assetDecimals), assetSymbol);
  console.log("");

  console.log("--- Direct price info ---");
  console.log("convertToAssets(", sharesInput, "):", ethers.formatUnits(grossAssets, assetDecimals), assetSymbol);
}

main().catch((error) => {
  console.error("Price/info flow failed:", error.shortMessage || error.message || error);
  process.exit(1);
});

File: examples/python-deposit-redeem-example/price.py
import os
import sys
import json
from dotenv import load_dotenv
from web3 import Web3

load_dotenv()


def get_env():
    rwa_token = os.getenv("RWA_TOKEN_ADDRESS")
    fee_address = os.getenv("FEE_ADDRESS") or ""

    env = {
        "RPC_URL": os.getenv("RPC_URL"),
        "RWA_TOKEN_ADDRESS": Web3.to_checksum_address(rwa_token) if rwa_token else "",
        "FEE_ADDRESS": Web3.to_checksum_address(fee_address) if fee_address else "",
        "ENTRY_FEE_TIMING": int(os.getenv("ENTRY_FEE_TIMING") or 0),
        "EXIT_FEE_TIMING": int(os.getenv("EXIT_FEE_TIMING") or 1),
    }

    if not all([env["RPC_URL"], env["RWA_TOKEN_ADDRESS"]]):
        raise ValueError("Missing required env values. Please check .env.example")

    return env


def load_abi(filename):
    path = os.path.join(os.path.dirname(__file__), "abis", filename)
    with open(path, "r") as f:
        data = json.load(f)
        return data.get("abi", data)


def resolve_fee_address(vault, env):
    if env["FEE_ADDRESS"]:
        return env["FEE_ADDRESS"]

    fee_address = vault.functions.rwaFee().call()
    if not fee_address or fee_address == "0x0000000000000000000000000000000000000000":
        raise ValueError(
            "Fee address is not configured. Set FEE_ADDRESS or ensure vault.rwaFee() returns a valid address."
        )

    return fee_address


def main():
    desired_assets_input = sys.argv[2] if len(sys.argv) > 2 else "100"
    shares_input = sys.argv[3] if len(sys.argv) > 3 else "25"

    env = get_env()

    w3 = Web3(Web3.HTTPProvider(env["RPC_URL"]))

    rwa_abi = load_abi("RWAToken.json")
    fee_abi = load_abi("RWAFee.json")

    vault = w3.eth.contract(address=env["RWA_TOKEN_ADDRESS"], abi=rwa_abi)

    asset_address = vault.functions.asset().call()

    asset = w3.eth.contract(
        address=asset_address,
        abi=[
            {
                "inputs": [],
                "name": "decimals",
                "outputs": [{"type": "uint8"}],
                "stateMutability": "view",
                "type": "function",
            },
            {
                "inputs": [],
                "name": "symbol",
                "outputs": [{"type": "string"}],
                "stateMutability": "view",
                "type": "function",
            },
        ],
    )

    share_decimals = vault.functions.decimals().call()
    share_symbol = vault.functions.symbol().call()
    asset_decimals = asset.functions.decimals().call()
    asset_symbol = asset.functions.symbol().call()

    fee_address = resolve_fee_address(vault, env)
    fee_contract = w3.eth.contract(address=fee_address, abi=fee_abi)

    desired_assets = int(float(desired_assets_input) * (10**asset_decimals))
    shares = int(float(shares_input) * (10**share_decimals))

    entry_fee = fee_contract.functions.feeOnRaw(
        desired_assets, env["ENTRY_FEE_TIMING"]
    ).call()
    total_spend = desired_assets + entry_fee
    preview_deposit_shares = vault.functions.previewDeposit(total_spend).call()

    gross_assets = vault.functions.convertToAssets(shares).call()
    estimated_exit_fee = fee_contract.functions.feeOnTotal(
        gross_assets, env["EXIT_FEE_TIMING"]
    ).call()
    preview_redeem_net_assets = vault.functions.previewRedeem(shares).call()

    print("=== PRICE / INFO FLOW ===")
    print("Vault:", env["RWA_TOKEN_ADDRESS"])
    print("Fee contract:", fee_address)
    print("")

    print("--- Deposit information ---")
    print("Desired assets before fee:", desired_assets_input, asset_symbol)
    print("Entry fee (feeOnRaw):", entry_fee / (10**asset_decimals), asset_symbol)
    print(
        "Total spend sent to deposit():",
        total_spend / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "previewDeposit(totalSpend):",
        preview_deposit_shares / (10**share_decimals),
        share_symbol,
    )
    print("")

    print("--- Redeem information ---")
    print("Input shares:", shares_input, share_symbol)
    print(
        "convertToAssets(shares) [gross info]:",
        gross_assets / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "Estimated exit fee (feeOnTotal):",
        estimated_exit_fee / (10**asset_decimals),
        asset_symbol,
    )
    print(
        "previewRedeem(shares) [final net]:",
        preview_redeem_net_assets / (10**asset_decimals),
        asset_symbol,
    )
    print("")

    print("--- Direct price info ---")
    print(
        "convertToAssets(",
        shares_input,
        "):",
        gross_assets / (10**asset_decimals),
        asset_symbol,
    )


if __name__ == "__main__":
    main()
